### ISSUE TITLE
Lua script will block grpc's stream

### ISSUE THREAD

**Issue by j2gg0s** [envoyproxy/envoy#24853]
*Description*:
We use [Lua](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/lua_filter#config-http-filters-lua-stream-handle-api) in HTTP Filters to log http request/response's body for debug, like this:
```lua
function envoy_on_request(handle)
  local headers = {}
  for k, v in pairs(handle:headers()) do
    headers[k] = v
  end
  handle:streamInfo():dynamicMetadata():set(
    "envoy.lua.log", "request_headers", headers)

  local md = handle:headers():get(":method")
  if md ~= "POST" and md ~= "PUT" then return end

  local body = handle:body(false)
  if (body ~= nil and body:length() > 0) then
    local n = body:length()
    if n >= 10240 then n = 10240 end

    if handle:headers():get("content-type") == "application/grpc" then
      handle:streamInfo():dynamicMetadata():set(
        "envoy.lua.log", "request_body", handle:base64Escape(body:getBytes(0, n)))
    else
      handle:streamInfo():dynamicMetadata():set(
        "envoy.lua.log", "request_body", body:getBytes(0, n))
    end
  end
end
```

This lua script work well with http/2 and grpc.
However, when it encounters grpc stream, all request will timeout.
This is beacause grpc's stream does not close http2's stream after one grpc's request,
so our lua script block and envoy dont send grpc's request to upstream.

My understanding is correct?
Such behaviour conform to our design?

**Comment by wbpcode** [comment]
`handle:body(false)` will wait for the end stream of HTTP body. When the grpc stream is used, there is no end stream, so the lua filter will blocks the filter chain processing.

**Comment by j2gg0s** [comment]
> `handle:body(false)` will wait for the end stream of HTTP body. When the grpc stream is used, there is no end stream, so the lua filter will blocks the filter chain processing.

@wbpcode thanks，but if i use `handle:body(true)`, i can't log http body?

**Comment by wbpcode** [comment]
Nope. `handle:body(true)` make no big difference with `handle:body(false)`. I think may only `bodyChunk` can help. The `bodyChunk` will stream the request and will not blocks the filter chain processing.

But I am still cann't ensure that everything of lua filter will be ok for grpc stream.

**Comment by j2gg0s** [comment]
Yes, `bodyChunk` will not block grpc stream. 
And we can exit follo loop when gprc stream end.
```
for chunk in handle:bodyChunk() do
end
```
